zeebe/engine/src/test/java/io/camunda/zeebe/engine/processing/processinstance/migration/MigrateErrorEventSubprocessTest.java

```diff
@@ -0,0 +1,351 @@
+        .describedAs("Expect that version number did not change")
+        .hasVersion(1);
+  }
+
+  @Test
+  public void shouldMigrateErrorEventForMappedStartEvent() {
+    // given
+    final String processId = helper.getBpmnProcessId();
+    final String targetProcessId = helper.getBpmnProcessId() + "2";
+    final var deployment =
+        ENGINE
+            .deployment()
+            .withXmlResource(
+                Bpmn.createExecutableProcess(processId)
+                    .eventSubProcess(
+                        "sub1",
+                        s ->
+                            s.startEvent("start1").error("error1").userTask("userTask1").endEvent())
+                    .startEvent("start")
+                    .serviceTask("A", t -> t.zeebeJobType("A"))
+                    .endEvent("end")
+                    .done())
+            .withXmlResource(
+                Bpmn.createExecutableProcess(targetProcessId)
+                    .eventSubProcess(
+                        "sub2",
+                        s ->
+                            s.startEvent("start2").error("error2").userTask("userTask2").endEvent())
+                    .startEvent("start")
+                    .serviceTask("B", t -> t.zeebeJobType("B"))
+                    .endEvent("end")
+                    .done())
+            .deploy();
+    final var processInstanceKey = ENGINE.processInstance().ofBpmnProcessId(processId).create();
+
+    RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_ACTIVATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .withElementId("A")
+        .await();
+
+    final long targetProcessDefinitionKey =
+        extractProcessDefinitionKeyByProcessId(deployment, targetProcessId);
+
+    // when
+    ENGINE
+        .processInstance()
+        .withInstanceKey(processInstanceKey)
+        .migration()
+        .withTargetProcessDefinitionKey(targetProcessDefinitionKey)
+        .addMappingInstruction("A", "B")
+        .addMappingInstruction("start1", "start2")
+        .migrate();
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords(ProcessInstanceIntent.ELEMENT_MIGRATED)
+                .withProcessInstanceKey(processInstanceKey)
+                .withElementType(BpmnElementType.SERVICE_TASK)
+                .getFirst()
+                .getValue())
+        .describedAs("Expect that process definition key is changed")
+        .hasProcessDefinitionKey(targetProcessDefinitionKey)
+        .describedAs("Expect that bpmn process id and element id changed")
+        .hasBpmnProcessId(targetProcessId)
+        .hasElementId("B")
+        .describedAs("Expect that version number did not change")
+        .hasVersion(1);
+
+    // note that the error code is changed since the error code is a static value set inside the
+    // element itself
+    ENGINE.job().ofInstance(processInstanceKey).withType("A").withErrorCode("error2").throwError();
```

Comment:
🔧 We need to clarify this to users to avoid surprises, I suggest one of:
- document the different behavior of errors compared to other event types like messages (static message name changes are left unchanged)
  - this kinda binds us to this behavior, as it may be hard to change later
- open a bug report to describe the different (unexpected) behavior
- open a bug report and highlight it in the docs

Reply:
I would go with the first option as we plan to support re-evaluating expressions on other elements.

Reply:
>I would go with the first option as we plan to support re-evaluating expressions on other elements.

@berkaycanbc How do you see that working?

Re-evaluating expressions will be something the user has to enable through the API (something like `.addMappingInstruction("source", "target", i -> i.evaluateMessageNameExpression())`). The default behavior for errors still differs from other event types. I've asked [here](https://camunda.slack.com/archives/CSQ2E3BT4/p1726043974128849) to see if someone knows why we don't just have error subscriptions.

Reply:
I concluded this one as I assumed re-evaluating expressions will be the default behaviour unless user specifies the otherwise. If it won't be the default behaviour, I think, we should fix it in this iteration to keep user experience consistent across element migrations.

Reply:
> I assumed re-evaluating expressions will be the default behaviour unless user specifies the otherwise

We've already [documented the behavior](https://docs.camunda.io/docs/components/concepts/process-instance-migration/#jobs-expressions-and-input-mappings), we can't/shouldn't change the behavior in a breaking manner. So, adding re-evaluation of expression will be something users have to specifically choose.

Reply:
Re-reading that docs, we could expand upon it clarifying that this only concerns the parts of the element that have already been used. Output mappings, and End-type execution listeners will use the new process definition.

Reply:
>I've asked [here](https://camunda.slack.com/archives/CSQ2E3BT4/p1726043974128849) to see if someone knows why we don't just have error subscriptions.

If we have a good reason that errors are truly different. Then, we can just document it as a special behavior to errors. Otherwise, I suggest we consider this a bug.

Reply:
Created a related issue here https://github.com/camunda/camunda/issues/22180